capsules/src/proximity.rs

```diff
@@ -0,0 +1,310 @@
+        // apps
+        let mut highest_lower_proximity: u8 = 0;
+        let mut lowest_upper_proximity: u8 = 255;
+
+        for cntr in self.apps.iter() {
+            cntr.enter(|app, _| {
+                if (app.lower_proximity > highest_lower_proximity) && app.subscribed {
+                    highest_lower_proximity = app.lower_proximity;
+                }
+                if (app.upper_proximity < lowest_upper_proximity) && app.subscribed {
+                    lowest_upper_proximity = app.upper_proximity;
+                }
+            });
+        }
+
+        // return values
+        Thresholds {
+            lower: highest_lower_proximity,
+            upper: lowest_upper_proximity,
+        }
+    }
+
+    fn call_driver(&self, command: ProximityCommand, arg1: usize, arg2: usize) -> ReturnCode {
+        match command {
+            ProximityCommand::ReadProximity => self.driver.read_proximity(),
+            ProximityCommand::ReadProximityOnInterrupt => self
+                .driver
+                .read_proximity_on_interrupt(arg1 as u8, arg2 as u8),
+            _ => ReturnCode::ENOSUPPORT,
+        }
+    }
+
+    fn configure_callback(&self, callback: Option<Callback>, app_id: AppId) -> ReturnCode {
+        self.apps
+            .enter(app_id, |app, _| {
+                app.callback = callback;
+                ReturnCode::SUCCESS
+            })
+            .unwrap_or_else(|err| err.into())
+    }
+}
+
+impl hil::sensors::ProximityClient for ProximitySensor<'_> {
+    fn callback(&self, temp_val: usize, command_type: usize) {
+        // Here we callback the values only to the apps which are relevant for the callback
+        // We also dequeue any command for a callback so as to remove it from the wait list and add other commands to continue
+        match command_type {
+            command_type if command_type == ProximityCommand::ReadProximity as usize => {
+                // Schedule callbacks for appropriate apps
+                for cntr in self.apps.iter() {
+                    cntr.enter(|app, _| {
+                        if app.subscribed
+                            && (command_type == (ProximityCommand::ReadProximity as usize))
+                        {
+                            app.callback.map(|mut cb| cb.schedule(temp_val, 0, 0));
+                            app.subscribed = false; // dequeue
+                        }
+                    });
+                }
+            }
+
+            command_type if command_type == ProximityCommand::ReadProximityOnInterrupt as usize => {
+                // Schedule callbacks for appropriate apps
+                for cntr in self.apps.iter() {
+                    cntr.enter(|app, _| {
+                        if app.subscribed
+                            && (command_type
+                                == (ProximityCommand::ReadProximityOnInterrupt as usize))
```

Comment:
If an app has requested an immediate reading, but the next reading to arrive is triggered by a proximity interrupt, might that app not still want to just receive that reading, rather than waiting for a second reading?

Reply:
I have addressed this.  Another good suggestion.